File: src/dotnet/shared/utilities/Result.cs
using System;
using System.Runtime.InteropServices;

namespace Hyperion
{
    // Maps to core/utilities/Result.hpp
    [ClassBinding(Name="Result")]
    [StructLayout(LayoutKind.Explicit, Size = 8)]
    public struct Result
    {
        [FieldOffset(0)]
        private PimplPtr _error;

        public bool IsValid
        {
            get
            {
                return _error.ptr == IntPtr.Zero;
            }
        }
    }
}

File: Source/Core/utilities/Result.cpp
#include <Core/utilities/Result.hpp>

#ifndef HYP_TOOL
#include <Result.generated.inl>
#endif

namespace Hyperion {
namespace utilities {

class NullError final : public Error
{
public:
    NullError() = default;
};

HYP_API const Error& GetNullError()
{
    static NullError s_nullError;

    return s_nullError;
}

} // namespace utilities
} // namespace Hyperion
